homeassistant/components/telegram_poll.py

```diff
@@ -0,0 +1,129 @@
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Setup is called when Home Assistant is loading our component."""
+    import telegram
+    bot_token = config[DOMAIN].get(BOT_TOKEN)
+    bot = telegram.Bot(bot_token)
+    return _setup(hass, config, bot)
+
+
+def _setup(hass, config, bot):
+    """The actual setup of the telegram component."""
+    allowed_chat_ids = config[DOMAIN].get(ALLOWED_CHAT_IDS)
+    checker = IncomingChecker(bot, hass, allowed_chat_ids)
+
+    def _start_bot(_event):
+        """Start the polling thread."""
+        checker.check_thread.start()
+
+    def _stop_bot(_event):
+        """Stop the thread."""
+        checker.checking = False
+
+    hass.bus.listen_once(
+        EVENT_HOMEASSISTANT_START,
+        _start_bot
+    )
+    hass.bus.listen_once(
+        EVENT_HOMEASSISTANT_STOP,
+        _stop_bot
+    )
+
+    # Return boolean to indicate that initialization was successfully.
+    return True
+
+
+class IncomingChecker:
+    """Threaded telegram incoming message handler."""
+
+    def __init__(self, bot, hass, allowed_ids):
+        """Initialize the thread."""
+        self.update_id = None
+        self.bot = bot
+        self.hass = hass
+        # boolean to check if checking loop should continue running
+        self.checking = True
+        self.allowed_ids = allowed_ids
+        self.check_thread = Thread(target=self.check_incoming)
+
+    def handle(self):
+        """" Receiving and processing incoming messages."""
+        for update in self.bot.getUpdates(offset=self.update_id, timeout=10):
+            self.update_id = update.update_id + 1
+
+            # Updates can also be empty.
+            if update.message:
+                # we only want to process text messages and coming from
+                # allowed ids
+                chat_id = str(update.message.chat_id)
+                if (update.message.text and
+                        chat_id in self.allowed_ids):
+                    command = update.message.text.split(' ')
+                    self.hass.bus.fire(EVENT_TELEGRAM_COMMAND, {
+                        ATTR_COMMAND: command[0],
+                        ATTR_ARGS: ' '.join(command[1:]),
+                        ATTR_FROM_FIRST: update.message.from_user.first_name,
+                        ATTR_FROM_LAST: update.message.from_user.last_name,
+                        ATTR_USER_ID: chat_id
+                    })
+
+    def check_incoming(self):
+        """"Loop which continuously checks for incoming telegram messages."""
+        # get the first pending update_id, this is so
+        # we can skip over it in case
+        # we get an "Unauthorized" exception.
+        import telegram
+        try:
+            self.update_id = self.bot.getUpdates()[0].update_id
+        except IndexError:
+            self.update_id = None
+        while self.checking:
+            try:
+                self.handle()
```

Comment:
@andrey-git I have been thinking about merging too, but then it should be renamed into something completely different. Also there will be little code overlap. But I can definitely see logic in your remark. What do you think ? Do a rename ?

The `getUpdates` method uses long polling with a timeout of 10 seconds. So I guess it won't stress CPU too much.